middleware/route_injector.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/gorilla/mux"
+)
+
+// RouteInjector is a middleware that injects the route name for the current request into the request context.
+//
+// The route name can be retrieved by calling ExtractRouteName.
+type RouteInjector struct {
+	RouteMatcher RouteMatcher
+}
+
+func (i RouteInjector) Wrap(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		routeName := getRouteName(i.RouteMatcher, r)
+		handler.ServeHTTP(w, WithRouteName(r, routeName))
+	})
+}
+
+// WithRouteName annotates r's context with the provided route name.
+//
+// The provided value must be suitable to use as a Prometheus label value.
+//
+// This method should generally only be used in tests: in production code, use RouteInjector instead.
+func WithRouteName(r *http.Request, routeName string) *http.Request {
+	ctx := context.WithValue(r.Context(), contextKeyRouteName, routeName)
+	return r.WithContext(ctx)
+}
+
+// ExtractRouteName returns the route name associated with this request.
+// This is the same route name used for trace and metric names, and is already suitable for use as a Prometheus label
+// value.
+func ExtractRouteName(r *http.Request) string {
```

Comment:
Nits:
* I'd suggest to take `context.Context` parameter only, since that's there we take the name from
* I'd also suggest to document that for this function to work, one must include `RouteInjector` middleware as one of the first middlewares, since that's what stores the route into the context.

Reply:
Addressed in 3996871